metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -0,0 +1,509 @@
+          bloopServer.connect(buildClient)
+        }
+        _ = {
+          cancelables.add(build)
+          buildServer.foreach(_.shutdown())
+          buildServer = Some(build)
+        }
+        _ <- build.initialize()
+        _ <- installWorkspaceBuildTargets(build)
+          .trackInStatusBar("$(sync) importing workspace")
+        _ = statusBar.addMessage("$(rocket) Imported workspace!")
+        _ <- compileSourceFiles(buffers.open.toSeq)
+      } yield BuildChange.Reconnected
+    }
+  }.recover {
+    case NonFatal(e) =>
+      val message =
+        "Failed to connect with build server, no functionality will work."
+      val details = " See logs for more details."
+      scribe.error(message, e)
+      languageClient.showMessage(
+        new MessageParams(MessageType.Error, message + details)
+      )
+      BuildChange.Failed
+  }
+
+  /**
+   * Visit every file and directory in the workspace and register
+   * toplevel definitions for scala source files.
+   */
+  private def indexWorkspaceScalaSources(): Future[Unit] = Future {
+    Files.walkFileTree(
+      workspace.toNIO,
+      new SimpleFileVisitor[Path] {
+        override def visitFile(
+            file: Path,
+            attrs: BasicFileAttributes
+        ): FileVisitResult = {
+          val path = AbsolutePath(file)
+          if (path.toLanguage.isScala) {
+            index.addSourceFile(path, None)
+          }
+          super.visitFile(file, attrs)
+        }
+        override def preVisitDirectory(
+            dir: Path,
+            attrs: BasicFileAttributes
+        ): FileVisitResult = {
+          val path = AbsolutePath(dir)
+          if (path.resolve("META-INF").isDirectory) {
+            FileVisitResult.SKIP_SUBTREE
+          } else if (dir.endsWith(".bloop") || dir.endsWith(".metals")) {
+            FileVisitResult.SKIP_SUBTREE
+          } else {
+            super.preVisitDirectory(dir, attrs)
+          }
+        }
+      }
+    )
+  }
+
+  private def timed[T](didWhat: String)(thunk: => Future[T]): Future[T] = {
+    val elapsed = new Timer(time)
+    val result = thunk
+    result.onComplete(_ => scribe.info(s"time: $didWhat in $elapsed"))
+    result
+  }
+
+  /**
+   * Index all build targets in the workspace.
+   */
+  private def installWorkspaceBuildTargets(
+      build: BuildServerConnection
+  ): Future[Unit] = timed("imported workspace") {
+    for {
+      workspaceBuildTargets <- build.server.workspaceBuildTargets().toScala
+      _ = buildTargets.addWorkspaceBuildTargets(workspaceBuildTargets)
+      _ = compilers.reset()
+      ids = workspaceBuildTargets.getTargets.map(_.getId)
+      scalacOptions <- build.server
+        .buildTargetScalacOptions(new ScalacOptionsParams(ids))
+        .toScala
+      _ = {
+        buildTargets.addScalacOptions(scalacOptions)
+        JdkSources().foreach(zip => index.addSourceJar(zip))
+      }
+      dependencySources <- build.server
```

Comment:
Break down into separate method, we can index dependencies and scalac options separately.